### ISSUE TITLE
How can I run Barracuda inference engine MRTK-WebRTC remote player stream?

### ISSUE THREAD

**Issue by Kabster17** [Unity-Technologies/barracuda-release#175]
Hello everyone, I have successfully created a mrtk-webrtc connection between my Hololens and my Desktop using Unity3D. 
Right now, I am trying to work on running Barracuda inference engine using yolo-onnx model on the Hololens stream in Unity itself.
Just would like to know, how can I go about doing so or if its even possible to begin with. I would really appreciate any help/guidance in this.

Thanks in advance!

**Comment by AlexRibard** [comment]
Hi! It should be!
You can checkout the doc here:
https://github.com/Unity-Technologies/barracuda-release/blob/release/1.0.0/Documentation%7E/index.md
In your case the steps is pretty easy:
* import barracuda package
* add the onnx model as a regular asset
* load the model/create a worker/input and execute said model
```
public NNModel model;
Model m_model;
IWorker m_worker;

void Start()
{
   m_model = ModelLoader.Load(model, false);
   m_worker = WorkerFactory.CreateWorker(WorkerFactory.Type.ComputePrecompiled, m_model, false);
}

void Update()
{
   Tensor input = new Tensor(1, 224, 224, 3)
   m_worker.Execute(input);
}
```

**Comment by Kabster17** [comment]
Hi, thanks for your reply. 

But for the case of yolo and in order to show the bounding boxes and labels of the detections how can I go about doing that?
And for the inputs, do I need to send in the texture data of the video stream or is there another way to do so?
I apologize if my questions are trivial/basic as I am really new Unity and Barracuda.

Based on other yolo-barracuda implementations they seem rather extensive in their scripts in which they have applied the inference engine mainly on the webcam texture but for the case of MRTK-WebRTC I am not quite sure.

Thanks.

**Comment by AlexRibard** [comment]
No problem at all!

* you can create a Tensor directly from a Texture, we have a constructor for that! 
```
var input = new Tensor(srcTexture: texture, channels: 3)
```
So usually you get the `WebCamTexture` (https://docs.unity3d.com/ScriptReference/WebCamTexture.html) and blit that to a RenderTexture which you use for the constructor. In your case, I am not too familiar with MRTK-WebRTC, but I assume you can get the video feed to a rendertexture somehow... (maybe ask on some other Unity forum)

* you are correct, yolo outputs the bounding boxes. This is very specific depending on how the yolo encodes said bounding boxes.
So once you have the result of the inference you need to reconstruct the bbox screen coordinates and then display them on screen.

I'll link you to a few good github from Keijiro Takahashi
https://github.com/keijiro/BlazePalmBarracuda
https://github.com/keijiro/BlazeFaceBarracuda
https://github.com/keijiro/IrisBarracuda

**Comment by Kabster17** [comment]
Hi, I have actually managed to get the video feed to a render texture by assigning the created render texture as the target texture for the camera which is projecting the quad which is streaming the HoloLens via the Web-RTC.

Not sure if its the right way to get the remote stream texture but its what I have managed to do recently. 
I am actually trying to adapt from these GitHub implementations in performing detections with tiny-yolo ONNX model:

https://github.com/keijiro/TinyYOLOv2Barracuda
https://github.com/wojciechp6/YOLO-UnityBarracuda

But they have been incorporating with WebCam texture mainly, can I ask if there is a way to rework it to use render texture instead?
Thanks a lot for your help!

**Comment by AlexRibard** [comment]
If you have a RenderTexture it's actually pretty straightforward:
```
   Tensor input = new Tensor(srcTexture:renderTexture, channels:3)
  ```
You can pass the render texture directly to the constructor and then pass the tensor as input to the worker
```
   m_worker.Execute(input);
```

**Comment by Kabster17** [comment]
Hi, so as per your suggestion I have attached this script:

using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using Unity.Barracuda;

public class remotestream : MonoBehaviour
{
    // Start is called before the first frame update
    public NNModel model;
    Model m_model;
    IWorker m_worker;
    public RenderTexture remote;

    void Start()
    {
        m_model = ModelLoader.Load(model, false);
        m_worker = WorkerFactory.CreateWorker(WorkerFactory.Type.ComputePrecompiled, m_model, false);
    }


    void Update()
    {
        Tensor input = new Tensor(srcTexture: remote, channels:3);
       m_worker.Execute(input);
        
    }
}

But I am not receiving any output tensors, am I doing anything wrong or missing anything?

**Comment by amirebrahimi** [comment]
Hi @Kabster17,

You need to get output back explicitly from the worker. Please check out:
https://docs.unity3d.com/Packages/com.unity.barracuda@1.4/manual/GettingStarted.html#fetching-outputs
https://docs.unity3d.com/Packages/com.unity.barracuda@1.4/manual/ModelOutput.html

Hope that helps.

**Comment by Kabster17** [comment]
Hi, can I ask if yolov3 is supported in Barracuda? 
I am trying to use custom YOLOv3 model trained with Darknet with Barracuda.
I managed to convert the model from Darknet to Keras and then to ONNX model but when importing it into Unity, the input shape of the model is not correct.

I am not sure why I am getting this as the input to the YOLO model
![image](https://user-images.githubusercontent.com/42634421/120972776-b0cf4300-c7a0-11eb-890c-323078c987af.png)

I appreciate some help in this, thanks.

**Comment by FlorentGuinier** [comment]
Hi @Kabster17,

We have various instance of yolov3 in our test suite, so without being able to promise every single version of yolov3 will work out of the box i think this should probably be fine, have you tried to run inference on it?

About the (0,0,0,3) shape: While this dimension notation can seems obscure, it is fine imho as -1 or 0 mean variable dimensions in term on input shape, you can verify that the onnx file actually have such shape using [netron](https://netron.app/), it is probably gonna be (0,3,0,0) then it get translated to Barracuda as variable N (batch), H & W (height and width) and 3 channels (RGB in that case probably).

Let me know if this make sense :)

Florent

**Comment by Kabster17** [comment]
Hi Florent, 

I apologize in advance for my misinterpretation or lack of knowledge in this area of development, please do bear with me. 
I am actually working on customizing this yolo-barracuda implementation:

https://github.com/wojciechp6/YOLO-UnityBarracuda

In which the author developed the yolo detections using tiny-yolov2 but for my customization I am planning to use YOLOv3 after importing the YOLOv3 model into Unity I get this error:
![image](https://user-images.githubusercontent.com/42634421/121140727-c316b300-c86c-11eb-8bf6-eae886304dfc.png)

Which is suggesting that the input width and height of the inputs have been detected to of 0 value. 

I am also attaching the model architectures of both the Tiny-v2 and custom YOLOv3 which I retrieved and saved from netron.
![YOLOv2 Tiny onnx](https://user-images.githubusercontent.com/42634421/121141069-23a5f000-c86d-11eb-80e2-c42be5ae3d0a.png)
![YOLOv3 onnx](https://user-images.githubusercontent.com/42634421/121141100-2dc7ee80-c86d-11eb-9cdf-d61232aa3b20.png)


Can I ask is there any issue with my conversion of the Darknet model to ONNX model?
Thanks in advance!

**Comment by FlorentGuinier** [comment]
Can you post both the code you are using to schedule the network and the onnx model please? 
Best way forward is a local repro imho.
PS: random weight are fine and you can also post via barracuda-support @ unity3d.com

**Comment by FlorentGuinier** [comment]
According to the error above my guess would be scheduling is creating a renderTexture from **model.Inputs[0].shape** witch are themselves 0. However would be much easyer with an actual repro.

**Comment by Kabster17** [comment]
Hi Florent, 

Here is link to my repo containing the scripts, adapted from https://github.com/wojciechp6/YOLO-UnityBarracuda. 
https://github.com/Kabster17/YOLO

I believe the issue is which the created renderTexture from model.Inputs[0].shape which is actually calling the TextureScaler class from WebCamDetector.cs but I am not sure why is the passing parameters are seen as 0.

**Comment by FlorentGuinier** [comment]
From what i can see 
`inputWidthHeight = nn.model.inputs[0].shape[1];` could be the root of the problem, at least i would start by looking here!

However by just looking at the repo it is hard to say, if you are stuck please create a _minimal_ repro project including the onnx file and share it there or via barracuda-support @ unity3d.com

**Comment by Kabster17** [comment]
Yes, I do too realise that the issue lies with inputWidthHeight = nn.model.inputs[0].shape[1] but I am not sure why the inputs are being read as 0 and not as a variable. 

I cant share my custom yolov3 model as its trained with client data but I did try the to incorporate the inference with a converted Tiny-YOLOv2 model and its producing the same behavior. 

Right now, I changed the input width and height to be fixed at 416 for the texture scaling and for the renderTexture. I am facing with the argument exception error as shown. 

![image](https://user-images.githubusercontent.com/42634421/121315436-eb6ce300-c93a-11eb-89b7-3aed4ce4e60c.png)

**Comment by FlorentGuinier** [comment]
> Yes, I do too realise that the issue lies with inputWidthHeight = nn.model.inputs[0].shape[1] but I am not sure why the inputs are being read as 0 and not as a variable.

nn.model.inputs[0].shape[1] simply return the value read from the network architecture, 0 or -1 means "variable"  size.

> Right now, I changed the input width and height to be fixed at 416 for the texture scaling and for the renderTexture. I am facing with the argument exception error as shown.

I still think that a minimal repro (in that case with your tiny-yolov2 model that you can share) is the way forward.  If privacy is an issue you can zero the weights and share via barracuda-support @ unity3d.com

**Comment by Kabster17** [comment]
Hi Florent,

I have created a simplified repro as linked: git@github.com:Kabster17/YOLO-Barracuda.git
Please open the sample scene in the scenes folder to access the scene in unity. 
The tiny-yolov2 converted model is the model which i have converted using Keras-onnx example method. 

I did do some changes to the code/scripts but would really appreciate if I can get the script to access the correct input size of the model. Please check the unity project. Thanks.

**Comment by amirebrahimi** [comment]
Hi @Kabster17,

I cloned your project and tried to run locally. I only have a webcam on my labtop, so not sure if that is the same setup you are using. I had overflow errors initially, but that is only because the Rects have x/y: NaN and width/height: Infinity, I just added a guard against that:

```C#
        Rect rect = box.rect;
        if (float.IsNaN(rect.x) || float.IsNaN(rect.y) || float.IsInfinity(rect.width) || float.IsInfinity(rect.height))
            return;
```

But although I don't get errors I also don't see any boxes. So, I investigated why that is. 

It goes back to previous comments made by @FlorentGuinier [here](https://github.com/Unity-Technologies/barracuda-release/issues/175#issuecomment-855947301) and [here](https://github.com/Unity-Technologies/barracuda-release/issues/175#issuecomment-856533838) and [here](https://github.com/Unity-Technologies/barracuda-release/issues/175#issuecomment-858387231). You cannot use the shape from the network to inform texture creation size if the network accepts dynamic input size. You have two options: (1) either use the Yolo with the fixed input sizes and use that to initialize or (2) use the Yolo w/ dynamic input size and initialize with the input you plan to provide. 

A hack I threw in just to have it run was in `YoloHandler.cs`:

```C#
    public List<ResultBox> Run(Texture2D tex)
    {
        // ...

        Tensor input = new Tensor(tex);
        inputWidthHeight = input.width;

        // ...
    }
```

But even after that there is a mess of boxes w/ varying labels that now show. I can assure you that the network is running, but unfortunately we don't have the resources to debug every user's individual project when it moves beyond the inference level. I can't speak to the software architecture of the project you have forked from, whether it is well designed, and whether it is robust.

**Comment by Kabster17** [comment]
Hi @amirebrahimi,

Thank you for getting back to me regarding this matter, yes I have also somehow managed to run the inference and the yolo model using the Barracuda implementation, and I managed to see multiple bounding boxes and labels showing as a mess like you have mentioned. 

I achieved this by specifying the texture scaler dimensions under the WebCamDetector.cs  and by changing the inputWidthHeight in the Yolohandler script as you have mentioned. Yes, I guess the network is running but the software architecture might not be supporting well with maybe converted custom weights (i am assuming) as the author was using Tiny-YOLOv2 model straight from the ONNX model zoo.

I have also made an attempt to change the model to used to my own custom converted YOLOv3 model but the inference produced was still a mess of labels but without bounding boxes being shown or displayed. But overflow errors were being generated, guess I will incorporate your method of adding the guard and continue to work on integrating the project. 

Thanks a lot for your help in this issue. I was really hoping for this to work out with a custom trained yolo model/weights : (
